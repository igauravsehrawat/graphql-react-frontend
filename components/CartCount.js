import React from 'react';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import styled from 'styled-components';

//
const AnimationStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
    backface-visibility: hidden;
  }
  /* Turns fill it and fill back*/
  .count-enter {
    transform: scale(4) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
`;

const Dot = styled.div`
  background: red;
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 1rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-size: 1rem;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

// react transition group
// Doubles up the DOT and then we can use transition
const CartCount = ({ count }) => (
  <AnimationStyles>
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        className="count"
        classNames="count"
        key={count}
        timeout={{ enter: 400, exit: 400 }}
      >
        <Dot>{count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  </AnimationStyles>
);

export default CartCount;
