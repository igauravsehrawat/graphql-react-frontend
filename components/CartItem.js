import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import formatMoney from '../lib/formatMoney';

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 2px solid ${props => props.theme.lightGrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
    width: 100px;
  }
  h3,
  p {
    margin: 0px;
  }
`;

const CartItem = ({ cartItem }) => (
  <CartItemStyles>
    <img alt={cartItem.item.description} src={cartItem.item.image} />
    <h3>{cartItem.item.title}</h3>
    <p>
      {formatMoney(cartItem.quantity * cartItem.item.price)}
      {' - '}
      {cartItem.quantity} &times; {formatMoney(cartItem.item.price)}
    </p>
  </CartItemStyles>
);

CartItem.propTypes = {
  cartItem: PropTypes.object.isRequired,
};

export default CartItem;
