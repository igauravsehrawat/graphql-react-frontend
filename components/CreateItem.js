import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Mutation } from 'react-apollo';
import Router from 'next/router';
import Form from './styles/Form';
import Error from './ErrorMessage';

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    price: 0,
    image: '',
    largeImage: '',
  };

  handleChange = e => {
    const { name, type, value } = e.target;
    // console.log({ name, value, type });
    const typeValue = type === 'number' ? parseFloat(value) : value;
    this.setState({
      [name]: typeValue,
    });
  };

  handleUpload = async e => {
    const { files } = e.target;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'devFits');
    const res = await fetch(
      'https://api.cloudinary.com/v1_1/af0lqrya/image/upload',
      {
        method: 'POST',
        body: data,
      }
    );
    const file = await res.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    });
  };

  render() {
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
            onSubmit={async e => {
              // prevent default
              e.preventDefault();
              // call mutation
              const res = await createItem();
              // console.log({ res });
              // change the route
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              });
            }}
          >
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="image">Image</label>
              <input
                id="image"
                name="image"
                type="file"
                onChange={this.handleUpload}
              />
              {this.state.image && (
                <img width="200" src={this.state.image} alt="Upload preview" />
              )}
              <label htmlFor="title">Title</label>
              <input
                id="title"
                name="title"
                type="text"
                value={this.state.title}
                placeholder="Enter title"
                required
                onChange={this.handleChange}
              />
              <label htmlFor="price">Price</label>
              <input
                id="price"
                name="price"
                type="number"
                value={this.state.price}
                placeholder="Enter price"
                required
                onChange={this.handleChange}
              />
              <label htmlFor="description"></label>
              <textarea
                id="description"
                name="description"
                value={this.state.description}
                placeholder="Enter description"
                required
                onChange={this.handleChange}
              />
              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
