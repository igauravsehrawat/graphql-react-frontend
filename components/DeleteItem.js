import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from './Items';

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

class DeleteItem extends Component {
  onUpdate = (cache, payload) => {
    // 1. Read cache
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    // 2. Filter cache
    data.items = data.items.filter(
      item => item.id !== payload.data.deleteItem.id
    );
    // 3. Write cache
    console.log(data);
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
  };

  render() {
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{ id: this.props.id }}
        update={this.onUpdate}
      >
        {(deleteItem, { loading, error }) => {
          {
            error && alert(error.message);
          }
          return (
            <button
              onClick={e => {
                if (confirm('Are you sure?')) {
                  deleteItem();
                }
              }}
            >
              {this.props.children}
            </button>
          );
        }}
      </Mutation>
    );
  }
}

export default DeleteItem;
