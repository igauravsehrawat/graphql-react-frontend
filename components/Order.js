import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query, Mutation } from 'react-apollo';
import PropTypes from 'prop-types';
import Head from 'next/head';
import { format } from 'date-fns';
import formatMoney from '../lib/formatMoney';
import OrderStyles from './styles/OrderStyles';
import Error from './ErrorMessage';

const SINGLE_ORDER_QUERY = gql`
  query order($id: ID!) {
    order(id: $id) {
      id
      charge
      total
      createdAt
      user {
        id
      }
      items {
        id
        title
        description
        image
        price
        quantity
      }
    }
  }
`;
export default class Order extends Component {
  render() {
    return (
      <Query
        query={SINGLE_ORDER_QUERY}
        variables={{
          id: this.props.id,
        }}
      >
        {({ data, error, loading }) => {
          if (error) return <Error error={error} />;
          if (loading) return <p>Loading...</p>;
          const { order } = data;
          return (
            <OrderStyles>
              <Head>
                <title>Dev Fits - Order ID: {order.id}</title>
              </Head>
              <p>
                <span>Order ID:</span>
                <span>{order.id}</span>
              </p>
              <p>
                <span>Date</span>
                <span>{format(order.createdAt, 'MMMM d, YYYY h:mm a')}</span>
              </p>
              <p>
                <span>Order total</span>
                <span>{formatMoney(order.total)}</span>
              </p>
              <p>
                <span>Item Count</span>
                <span>{order.items.length}</span>
              </p>
              <div className="items">
                {order.items.map(item => (
                  <div key={item.id} classNames="order-item">
                    <img src={item.image} alt={item.description} />
                    <div classNames="item-details" key={item.id}>
                      <h2>{item.title}</h2>
                      <p>Qty: {item.quantity}</p>
                      <p>Each: {item.price}</p>
                      <p>
                        Sub Total: {formatMoney(item.price * item.quantity)}
                      </p>
                      <p>{item.description}</p>
                    </div>
                  </div>
                ))}
              </div>
            </OrderStyles>
          );
        }}
      </Query>
    );
  }
}

Order.propTypes = {
  id: PropTypes.string.isRequired,
};

export { SINGLE_ORDER_QUERY };
