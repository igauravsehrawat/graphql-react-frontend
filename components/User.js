import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

const CURRENT_USER_QUERY = gql`
  query {
    me {
      id
      name
      email
      permissions
      orders {
        id
      }
      cart {
        id
        quantity
        item {
          id
          title
          price
          description
          image
        }
      }
    }
  }
`;

const User = props => (
  <Query {...props} query={CURRENT_USER_QUERY}>
    {payload => props.children(payload)}
  </Query>
);

User.propTypes = {
  children: PropTypes.func.isRequired,
};

export default User;
export { CURRENT_USER_QUERY };
