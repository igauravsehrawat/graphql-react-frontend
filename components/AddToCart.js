import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query, Mutation } from 'react-apollo';
import { CURRENT_USER_QUERY } from './User';

const ADD_TO_CART_MUTATION = gql`
  mutation addToCart($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`;

export default class AddToCart extends Component {
  render(props) {
    const { id } = this.props;
    return (
      <Mutation
        mutation={ADD_TO_CART_MUTATION}
        variables={{
          id,
        }}
        refetchQueries={[{ query: CURRENT_USER_QUERY }]}
      >
        {(addToCart, { loading }) => (
          <button onClick={addToCart} disabled={loading}>
            Add{loading && 'ing'} To Cart ðŸ›’
          </button>
        )}
      </Mutation>
    );
  }
}

export { ADD_TO_CART_MUTATION };
