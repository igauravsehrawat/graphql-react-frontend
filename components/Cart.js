import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query, Mutation } from 'react-apollo';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import CartStyles from './styles/CartStyles';
import SickButton from './styles/SickButton';
import { CURRENT_USER_QUERY } from './User';

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation toggleCart {
    toggleCart @client {
      cartOpen
    }
  }
`;

export default class Cart extends Component {
  render(props) {
    return (
      <Query query={CURRENT_USER_QUERY}>
        {({ data: { me }, loading, error }) => (
          <Mutation mutation={TOGGLE_CART_MUTATION}>
            {toggleCart => (
              <Query query={LOCAL_STATE_QUERY}>
                {({ data, loading, error }) => {
                  {
                    console.log('data is...', data, me);
                  }
                  return (
                    <CartStyles open={data.cartOpen}>
                      <header>
                        <CloseButton title="close" onClick={toggleCart}>
                          &times;
                        </CloseButton>
                        <Supreme>{me.name}'s Cart</Supreme>
                        <p>
                          You have {me.cart.length} item{' '}
                          {me.cart.length > 1 && 's'} in your cart.
                        </p>
                        <ul>
                          {me.cart.map(item => (
                            <li>{item.id}</li>
                          ))}
                        </ul>
                      </header>
                      <footer>
                        <p>Your total is</p>
                        <SickButton>Checkout</SickButton>
                      </footer>
                    </CartStyles>
                  );
                }}
              </Query>
            )}
          </Mutation>
        )}
      </Query>
    );
  }
}
