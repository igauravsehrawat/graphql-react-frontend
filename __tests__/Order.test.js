import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import NProgress from 'nprogress';
import Router from 'next/router';
import { ApolloConsumer } from 'react-apollo';
import { MockedProvider } from 'react-apollo/test-utils';
import Order, { SINGLE_ORDER_QUERY } from '../components/Order';
import { fakeUser, fakeOrder } from '../lib/testUtils';
import { CURRENT_USER_QUERY } from '../components/User';

const mocks = [
  {
    request: {
      query: SINGLE_ORDER_QUERY,
      variables: { id: 'ord123' },
    },
    result: {
      data: {
        order: fakeOrder(),
      },
    },
  },
];

describe('<Order />', () => {
  it('renders and matches the snap shot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Order id="ord123" />
      </MockedProvider>
    );
    // await wait is when data is to rendered
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('OrderStyles'))).toMatchSnapshot();
  });
});
