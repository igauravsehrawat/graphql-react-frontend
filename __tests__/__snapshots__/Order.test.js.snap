// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order /> renders and matches the snap shot 1`] = `
<OrderStyles>
  <div
    className="OrderStyles-sc-4oqalm-0 camMij"
  >
    <SideEffect(Head)>
      <Head />
    </SideEffect(Head)>
    <p>
      <span>
        Order ID:
      </span>
      <span>
        ord123
      </span>
    </p>
    <p>
      <span>
        Date
      </span>
      <span>
        April 1, 2018 2:00 AM
      </span>
    </p>
    <p>
      <span>
        Order total
      </span>
      <span>
        €400
      </span>
    </p>
    <p>
      <span>
        Item Count
      </span>
      <span>
        2
      </span>
    </p>
    <div
      className="items"
    >
      <div
        classNames="order-item"
        key="3a430a73-c9e9-4bcc-b46a-41618965ffea"
      >
        <img
          alt="et modi tenetur amet modi reprehenderit omnis"
          src="reprehenderit.jpg"
        />
        <div
          classNames="item-details"
          key="3a430a73-c9e9-4bcc-b46a-41618965ffea"
        >
          <h2>
            soluta non omnis consequatur enim quia autem
          </h2>
          <p>
            Qty: 
            1
          </p>
          <p>
            Each: 
            4234
          </p>
          <p>
            Sub Total: 
            €42.34
          </p>
          <p>
            et modi tenetur amet modi reprehenderit omnis
          </p>
        </div>
      </div>
      <div
        classNames="order-item"
        key="2393e090-592f-4d03-b808-c9d81098deec"
      >
        <img
          alt="sapiente laudantium molestias assumenda quasi adipisci mollitia"
          src="sint.jpg"
        />
        <div
          classNames="item-details"
          key="2393e090-592f-4d03-b808-c9d81098deec"
        >
          <h2>
            quia sed exercitationem omnis laborum exercitationem est
          </h2>
          <p>
            Qty: 
            1
          </p>
          <p>
            Each: 
            4234
          </p>
          <p>
            Sub Total: 
            €42.34
          </p>
          <p>
            sapiente laudantium molestias assumenda quasi adipisci mollitia
          </p>
        </div>
      </div>
    </div>
  </div>
</OrderStyles>
`;
